### PR TITLE
virtcontainers: fix hotplug block/net devices execeed pciBridgeMaxCap…

### PR DESCRIPTION
**Pull request by woshijpf**
…acity bug

add rollback operations when hotplug block/net devices execeed pciBridgeMaxCapacity

Fixes: #1941

Signed-off-by: jiangpengfei <jiangpengfei9@huawei.com>

### CODE REVIEW

File: virtcontainers/qemu.go [hunk 1]
```diff
@@ -1012,6 +1010,12 @@ func (q *qemu) hotplugAddBlockDevice(drive *config.BlockDrive, op operation, dev
...
 
+	defer func() {
+		if err != nil {
+			q.qmpMonitorCh.qmp.ExecuteBlockdevDel(q.qmpMonitorCh.ctx, drive.ID)
+		}
+	}()
+
 	if q.config.BlockDeviceDriver == config.VirtioBlock {
 		driver := "virtio-blk-pci"
 		addr, bridge, err := q.addDeviceToBridge(drive.ID)
```
Comment (devimc) [inline virtcontainers/qemu.go]: Good catch!  I think `removeDeviceFromBridge` should be called if `ExecutePCIDeviceAdd` fails, right?

File: virtcontainers/qemu.go [hunk 2]
```diff
@@ -1173,6 +1177,12 @@ func (q *qemu) hotplugNetDevice(endpoint Endpoint, op operation) error {
 			return err
 		}
 
+		defer func() {
+			if err != nil {
+				q.qmpMonitorCh.qmp.ExecuteNetdevDel(q.qmpMonitorCh.ctx, tap.Name)
+			}
+		}()
+
 		addr, bridge, err := q.addDeviceToBridge(tap.ID)
```
Comment (devimc) [inline virtcontainers/qemu.go]: and same here ...

File: virtcontainers/qemu.go (hunk 1 above)
Comment (woshijpf) [reply virtcontainers/qemu.go]: You are right! `ExecutePCIDeviceAdd` fails need to rollback too.

File: virtcontainers/qemu.go [hunk 3]
```diff
@@ -1012,13 +1010,23 @@ func (q *qemu) hotplugAddBlockDevice(drive *config.BlockDrive, op operation, dev
...
+
 	if q.config.BlockDeviceDriver == config.VirtioBlock {
 		driver := "virtio-blk-pci"
 		addr, bridge, err := q.addDeviceToBridge(drive.ID)
 		if err != nil {
 			return err
 		}
 
+		defer func() {
+			q.removeDeviceFromBridge(drive.ID)
```
Comment (devimc) [inline virtcontainers/qemu.go]: missing `if err != nil {` ?

**Comment by GabyCT** [comment]
@woshijpf travis is failing mainly because of this
```
ERROR: Commit 371fc54ea9e2fd4cc3724df238f3af8c0e37c42a: subject too long (max 75, got 78): "virtcontainers: fix hotplug block/net devices execeed pciBridgeMaxCapacity bug"
```
could you please fix this? thanks

File: virtcontainers/qemu.go (hunk 3 above)
Comment (woshijpf) [reply virtcontainers/qemu.go]: 😅sorry for miss this check.

**Comment by woshijpf** [comment]
@GabyCT Thanks for you remind I have fixed this error.

**Comment by amshinde** [comment]
Rerunning jenkins-ci-ubuntu-nemu CI.

**Pull request merged**